igor-web/src/main/groovy/com/netflix/spinnaker/igor/docker/DockerMonitor.groovy

```diff
@@ -50,73 +64,72 @@ class DockerMonitor extends CommonPollingMonitor {
         dockerRegistryAccounts.accounts.forEach({ account ->
-            changedTags(account)
+            internalPoll(new PollContext((String) account.name, account))
         })
     }
 
-    private void changedTags(Map accountDetails) {
-        String account = accountDetails.name
-        Boolean trackDigests = accountDetails.trackDigests ?: false
-
-        log.debug 'Checking for new tags for ' + account
-        try {
-            List<String> cachedImages = cache.getImages(account)
-
-            def startTime = System.currentTimeMillis()
-            List<TaggedImage> images = dockerRegistryAccounts.service.getImagesByAccount(account)
-            log.debug("Took ${System.currentTimeMillis() - startTime}ms to retrieve images (account: {})", kv("account", account))
-
-            Map<String, TaggedImage> imageIds = images.collectEntries {
-                [(cache.makeKey(account, it.registry, it.repository, it.tag)): it]
+    @Override
+    DockerPollingDelta generateDelta(PollContext ctx) {
+        String account = ctx.context.name
+        Boolean trackDigests = ctx.context.trackDigests ?: false
+
+        log.debug("Checking new tags for {}", account)
+        List<String> cachedImages = cache.getImages(account)
+
+        long startTime = System.currentTimeMillis()
+        List<TaggedImage> images = dockerRegistryAccounts.service.getImagesByAccount(account)
+        log.debug("Took ${System.currentTimeMillis() - startTime}ms to retrieve images (account: {})", kv("account", account))
+
+        // Find the delta of cache items
+        List<ImageDelta> delta = []
+        images.parallelStream().forEach({ TaggedImage image ->
+            String imageId = cache.makeKey(account, image.registry, image.repository, image.tag)
+            boolean updateCache = false
+
+            if (imageId in cachedImages) {
+                if (trackDigests) {
+                    String lastDigest = cache.getLastDigest(image.account, image.registry, image.repository, image.tag)
+
+                    if (lastDigest != image.digest) {
+                        log.info("Updated tagged image: {}: {}. Digest changed from [$lastDigest] -> [$image.digest].", kv("account", image.account), kv("image", imageId))
+                        // If either is null, there was an error retrieving the manifest in this or the previous cache cycle.
+                        updateCache = image.digest != null && lastDigest != null
+                    }
+                }
+            } else {
+                updateCache = true
             }
 
-            /* Not removing images from igor as we're seeing some reading issues in clouddriver
-            Observable.from(cachedImages).filter { String id ->
-                !(id in imageIds)
-            }.subscribe({ String imageId ->
-                log.info "Removing $imageId."
-                cache.remove(imageId)
-            }, {
-                log.error("Error: ${it.message}")
+            if (updateCache) {
```

Comment:
I would pull the computation for updateCache in its own function, something like this:

```
def updateCache() {
    if (!imageId in cachedImages) {
        return true
    }

    if (trackDigests) {
        String lastDigest = cache.getLastDigest(image.account, image.registry, image.repository, image.tag)

        if (lastDigest != image.digest) {
            log.info("Updated tagged image: {}: {}. Digest changed from [$lastDigest] -> [$image.digest].", kv("account", image.account), kv("image", imageId))
            // If either is null, there was an error retrieving the manifest in this or the previous cache cycle.
            return image.digest != null && lastDigest != null
        }
    }

    return false
}
```

and then you don't need the variable I think:

```
if (updateCache()) {
    delta.add(new ImageDelta(imageId: imageId, image: image))
}
```

Reply:
I would also consider renaming the function, as it does not "update the cache". Maybe something like `shouldUpdateCache`?

Reply:
👍 to this suggestion